HeterogeneousCore/CUDACore/interface/CUDAContextState.h

```diff
@@ -0,0 +1,57 @@
+#ifndef HeterogeneousCore_CUDACore_CUDAContextState_h
+#define HeterogeneousCore_CUDACore_CUDAContextState_h
+
+#include "HeterogeneousCore/CUDAUtilities/interface/SharedStreamPtr.h"
+
+#include <memory>
+
+/**
+ * The purpose of this class is to deliver the device and CUDA stream
+ * information from ExternalWork's acquire() to producer() via a
+ * member/StreamCache variable.
+ */
+class CUDAContextState {
```

Comment:
Would the use of a namespace be a good idea for code in this package?

Reply:
> Would the use of a namespace be a good idea for code in this package?

Possibly. The different naming convention in `CUDACore` (`CUDA` prefix) and in `CUDAUtilities` (`cudautils` namespace) has disturbed me a bit. The question is then what the namespace should be?

Earlier the short `cuda` was "taken" by the API wrappers, but we have removed its use. Alternative would be `cudacore`. Does anyone have better suggestions?

Reply:
Or `edm` and keep the `CUDA` prefix?

Reply:
About namespaces, I was thinking we could use three different ones:
  - `cudautils` (or just `cuda` ?) for CUDA specific utilities, that are not dependent on CMS code;
  - `cudacompat` for the utilities, helpers, etc., that are not dependent on CMS code, that we use to compile the CUDA-like code for the host (these tools can use or reimplement the previous ones);
  - something else, for the EDM interface to CUDA; could be `cudacore::Whatever`, `edm::CUDAWhatever`, `cms::cuda::Whatever`, `edm::cuda::Whatever`,  ...

I haven't been able to come up with better names :-(

Reply:
With @Dr15Jones we did a quick search on google and github, and it seems that `cudautils` seems to be a common namespace in various software packages (with various capitalization conventions). Therefore we're thinking to go with `cms::cuda::Whatever` for everything in `CUDACore` and `CUDAUtilities`.

I'm thinking to keep the `cudacompat` as separate (`cms::cudacompat`) as it is very specific case and has tricky contents. (same for `cudatest` -> `cms::cudatest`).

Reply:
I would really prefer avoid having CUDA-related utilities that are not depending or designed for CMSSW inside `namespace cms`:
  - `cudaCheck` (ok, that's a macro, so no namespace for it)
  - `cudautils::launch`
  - the allocators and memory operations
  - _etc._

If `cudautils` is overused, let's pick something else, like `cudax::` or `cudaex::` (I quick search on GitHujb didn't reveal much use of those) or even `patatrack::` .

Reply:
The main motivation for `namespace cms` would be the code being inside CMSSW, and `cms` being "our unique name" (and yes, overall we don't really use namespaces consistently or well) would be the best protection against name collisions.

After that the argumentation goes with consistency among CUDA tools: `cuda<something>` is risky for some values of `<something>`, so if one goes with `cms::` prefix with one, it would be consistent to use it for all. (why have a mixture of very different namespaces for CUDA tools that a developer is supposed to interact with?)

Where would the line be in whether something depends or not on CMSSW? For example,
* `cudaCheck.h` uses `FWCore/Utilities/interface/Likely.h`
   * In the future it may have to throw the exceptions as `edm::Exception` in case computing wants specific error code for CUDA errors
* In the allocators `Caching*Allocator.h` and `allocate_*.h` themselves are independent of CMSSW, but
  * `allocate_*.cc` and `*_unique_ptr.h` use `FWCore/Utilities/interface/Likely.h`
  * `getCaching*Allocator.h` use `FWCore/MessageLogger/interface/MessageLogger.h` and `FWCore/Utilities/interface/thread_safety_macros.h`
  * future rewrite of the allocators might well use MessageLogger (unless specifically intended for wider use)
* Is the "dependence on CMSSW" a transitive property?
   * For example with the memory operations
      * `getCachingDeviceAllocator.h` depends on MessageLogger, so let's say it "depends on CMSSW"
      * `allocate_device.cc` depends on `getCachingDeviceAllocator.h`, do the functions in there now "depend on CMSSW" as well?
      * If they do, `device_unique_ptr.h` depends on `allocate_device.h`, does it "depend on CMSSW" as well?
      * If it does, `copyAsync.h` and `memsetAsync.h` depend on `device_unique_ptr.h`, do they "depend on CMSSW" as well?

I don't have a strong opinion what the exact name(s) should be, but I'd like to have a solution that is future-proof(-enough), consistent(-enough), and convenient(-enough).

Reply:
> The main motivation for `namespace cms` would be the code being inside CMSSW, and `cms` being "our unique name" (and yes, overall we don't really use namespaces consistently or well) would be the best protection against name collisions.

I think that could make sense in theory, though I would hate to see all the types and functions in our code prefixed with `cms::`... 

> After that the argumentation goes with consistency among CUDA tools: `cuda<something>` is risky for some values of `<something>`, so if one goes with `cms::` prefix with one, it would be consistent to use it for all. (why have a mixture of very different namespaces for CUDA tools that a developer is supposed to interact with?)

You mean like having to deal with `edm::`, `cms::`, `reco::`, etc. ?
We even have `edm::Exception` and `cms::Exception` :-(


> Where would the line be in whether something depends or not on CMSSW? For example,
> 
>     * `cudaCheck.h` uses `FWCore/Utilities/interface/Likely.h`

Well, that's because you insisted on using it.
Initially is used `__builtin_expect()` instead, with the same effect minus the dependency on the FWCore/Utilities package.

>       
>       * In the future it may have to throw the exceptions as `edm::Exception` in case computing wants specific error code for CUDA errors

IMHO we should do it the other way around: if we want to use more specific exceptions, we should use `std::error_code` with a new error category (e.g. `cuda_error` inheriting from `std::error_category`); if the framework needs to report all cuda errors with a specific error code, it could catch all `cuda_error` exceptions and react accordingly.

>     * In the allocators `Caching*Allocator.h` and `allocate_*.h` themselves are independent of CMSSW, but
>       
>       * `allocate_*.cc` and `*_unique_ptr.h` use `FWCore/Utilities/interface/Likely.h`
>       * `getCaching*Allocator.h` use `FWCore/MessageLogger/interface/MessageLogger.h` and `FWCore/Utilities/interface/thread_safety_macros.h`
>       * future rewrite of the allocators might well use MessageLogger (unless specifically intended for wider use)
>
>     * Is the "dependence on CMSSW" a transitive property?

Yes.
   
>       * For example with the memory operations
>         
>         * `getCachingDeviceAllocator.h` depends on MessageLogger, so let's say it "depends on CMSSW"
>         * `allocate_device.cc` depends on `getCachingDeviceAllocator.h`, do the functions in there now "depend on CMSSW" as well?
>         * If they do, `device_unique_ptr.h` depends on `allocate_device.h`, does it "depend on CMSSW" as well?
>         * If it does, `copyAsync.h` and `memsetAsync.h` depend on `device_unique_ptr.h`, do they "depend on CMSSW" as well?

So the whole chain points to `MessageLogger`, which is indeed a rather intrusive thing... I would not consider it "CMS specific", in the sense that we could work around it if needed (for example with a macro that expands to `std::cerr << category << ": "` by default, and to the `LogError(category)` in CMSSW).


> I don't have a strong opinion what the exact name(s) should be, but I'd like to have a solution that is future-proof(-enough), consistent(-enough), and convenient(-enough).

In my mind the distinction is mainly among three classes:
  1. utilities like memory allocation, copies or kernel launches - which kind of extend the CUDA API
  2. utilities that we currently use to compile CUDA-like code on the host
  3. classes like the CUDA dataformts, context, event and stream caches, that we use to wrap CUDA and interface it with the framework

Are you suggesting that we use
  - a single namespace (e.g. `cms::cuda::`) for all of them
  - one namespace (e.g. `cms::cuda::`) for 1. and 3., and a separate on (e.g. `cms::cuda_compat::` or `cms::cuda::compat::`) for 2.
  - a different namespaces for each one (e.g. `cms::cuda::`,  `cms::cuda_compat:::` and `edm::`)
?

Reply:
> You mean like having to deal with `edm::`, `cms::`, `reco::`, etc. ?

Not really the need to deal with various namespaces, but things are not always put to these (or other) namespaces consistently.

> We even have `edm::Exception` and `cms::Exception` :-(

Indeed...

> IMHO we should do it the other way around: if we want to use more specific exceptions, we should use `std::error_code` with a new error category (e.g. `cuda_error` inheriting from `std::error_category`); if the framework needs to report all cuda errors with a specific error code, it could catch all `cuda_error` exceptions and react accordingly.

Where would such `cuda_error` be defined? If it would be outside of `FWCore`, `FWCore` should not be made dependent on it. If inside `FWCore`, what would be the benefit over just using `edm::Exception`?

Anyway, the exception type was purely speculative at this point, i.e. for now `std::runtime_error` is fine.

But I think it would be good to write down some justification for "why this piece of code in CMSSW should try hard to avoid the dependence on the framework?"

> In my mind the distinction is mainly among three classes:
> 
> 1. utilities like memory allocation, copies or kernel launches - which kind of extend the CUDA API
> 2. utilities that we currently use to compile CUDA-like code on the host
> 3. classes like the CUDA dataformts, context, event and stream caches, that we use to wrap CUDA and interface it with the framework

I'd add "4. classes/functions for tests".

> * one namespace (e.g. `cms::cuda::`) for 1. and 3., and a separate on (e.g. `cms::cuda_compat::` or `cms::cuda::compat::`) for 2.

This is the one I was thinking of (with `cms::cudacompat`, could be either one of above as well; and `cms::cudatest` for 4.). I think 2.  is special-enough to motivate a separate namespace. I'm less sure of the benefits to distinguish 1. and 3. at the namespace level.

From the other way around, 3. should not really be part of `edm` (they are not components of the framework itself). Their namespace should not be just `cuda`. The `cms::cuda` would be ok. With that `cudautils` for 1. feels a bit inconsistent, `cms::cuda` (or `cms::cudautils` if 1. and 3. would be beneficial to separate at namespace level). In fact in some sense `cms::cuda` as a name would fit better for 1. than for 3. ("extending CUDA API").

Or we accept the risk of possible future name collisions ("to be dealt with later") and stay with `cudautils` for 1., in which case 3. would likely get `cudacore` (or how bad would `cudautils` sound for it?).

Reply:
To summarize the discussion from the core meeting, I'll proceed with
1. (utilities): `cms::cuda` namespace
2. (utilities to compile CUDA-like code on CPU): `cms::cudacompat` namespace (or similar, not part of this PR)
3. (utilities to interact with framework): `cms::cuda` namespace
4. (test-related stuff): `cms::cudatest` namespace